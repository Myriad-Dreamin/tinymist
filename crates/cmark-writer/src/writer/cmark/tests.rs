use super::CommonMarkWriter;
use crate::ast::{HeadingType, Node};
use crate::options::WriterOptions;
use crate::writer::runtime::diagnostics::{DiagnosticSeverity, SharedVecSink};
use std::cell::RefCell;
use std::rc::Rc;

fn render(node: &Node) -> String {
    let mut writer = CommonMarkWriter::new();
    writer.write(node).unwrap();
    writer.into_string().into()
}

#[test]
fn paragraphs_are_separated_by_blank_line() {
    let document = Node::Document(vec![
        Node::Paragraph(vec![Node::Text("First".into())]),
        Node::Paragraph(vec![Node::Text("Second".into())]),
    ]);

    assert_eq!(render(&document), "First\n\nSecond\n");
}

#[test]
fn setext_heading_matches_content_width() {
    let heading = Node::Heading {
        level: 2,
        content: vec![Node::Text("Wide Title".into())],
        heading_type: HeadingType::Setext,
    };

    assert_eq!(render(&heading), "Wide Title\n----------\n");
}

#[test]
fn autolink_preserves_url() {
    let autolink = Node::Autolink {
        url: "example.com/path".into(),
        is_email: false,
    };

    assert_eq!(render(&autolink), "<example.com/path>");
}

#[test]
fn thematic_break_ends_with_newline() {
    let document = Node::Document(vec![
        Node::ThematicBreak,
        Node::Paragraph(vec![Node::Text("After".into())]),
    ]);

    assert_eq!(render(&document), "---\n\nAfter\n");
}

#[test]
fn diagnostics_capture_warnings() {
    let options = WriterOptions {
        strict: false,
        ..WriterOptions::default()
    };

    let mut writer = CommonMarkWriter::with_options(options);
    let storage = Rc::new(RefCell::new(Vec::new()));
    let sink = SharedVecSink::new(storage.clone());
    writer.set_diagnostic_sink(Box::new(sink));

    let autolink = Node::Autolink {
        url: "foo\nbar".into(),
        is_email: false,
    };

    writer.write(&autolink).unwrap();

    let diagnostics = storage.borrow();
    assert_eq!(diagnostics.len(), 1);
    assert_eq!(diagnostics[0].severity, DiagnosticSeverity::Warning);
    assert!(diagnostics[0]
        .message
        .contains("Newline character found in autolink URL"));
}
