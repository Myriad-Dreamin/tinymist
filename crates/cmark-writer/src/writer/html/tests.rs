use crate::ast::{HtmlAttribute, HtmlElement, Node};
use crate::{HtmlWriter, HtmlWriterOptions};

#[test]
fn write_trusted_html_keeps_fragment_verbatim() {
    let mut writer = HtmlWriter::new();
    writer.start_tag("div").unwrap();
    writer.finish_tag().unwrap();
    writer.write_trusted_html("<span>").unwrap();
    writer.write_trusted_html("&ok").unwrap();
    writer.end_tag("div").unwrap();

    let output = writer.into_string().unwrap();
    assert_eq!(output, "<div><span>&ok</div>");
}

#[test]
fn write_untrusted_html_escapes_fragment() {
    let mut writer = HtmlWriter::new();
    writer.start_tag("div").unwrap();
    writer.finish_tag().unwrap();
    writer.write_untrusted_html("<span>&oops").unwrap();
    writer.end_tag("div").unwrap();

    let output = writer.into_string().unwrap();
    assert_eq!(output, "<div>&lt;span&gt;&amp;oops</div>");
}

#[test]
fn attribute_escaping_handles_quotes_and_special_chars() {
    let mut writer = HtmlWriter::new();
    writer.start_tag("div").unwrap();
    writer
        .attribute("data-title", "He said \"<Hello>\" & more")
        .unwrap();
    writer.finish_self_closing_tag().unwrap();

    let output = writer.into_string().unwrap();
    assert_eq!(
        output,
        "<div data-title=\"He said &quot;&lt;Hello&gt;&quot; &amp; more\" />"
    );
}

#[test]
fn guarded_writer_renders_safe_html_element() {
    let mut writer = HtmlWriter::new();
    let element = HtmlElement {
        tag: "div".into(),
        attributes: vec![HtmlAttribute {
            name: "class".into(),
            value: "note".into(),
        }],
        children: vec![Node::Text("Hello".into())],
        self_closing: false,
    };

    writer.write_html_element(&element).unwrap();

    let output = writer.into_string().unwrap();
    assert_eq!(output, "<div class=\"note\">\n\nHello\n\n</div>");
}

#[test]
fn guarded_writer_textualizes_invalid_tag_in_non_strict_mode() {
    let mut writer = HtmlWriter::with_options(HtmlWriterOptions {
        strict: false,
        ..Default::default()
    });
    let element = HtmlElement {
        tag: "div!".into(),
        attributes: vec![HtmlAttribute {
            name: "class".into(),
            value: "unsafe".into(),
        }],
        children: vec![Node::Text("oops".into())],
        self_closing: false,
    };

    writer.write_html_element(&element).unwrap();

    let output = writer.into_string().unwrap();
    assert_eq!(
        output,
        "&lt;div! class=\"unsafe\"&gt;\n\noops\n\n&lt;/div!&gt;"
    );
}

#[test]
fn guarded_writer_errors_on_invalid_tag_in_strict_mode() {
    let mut writer = HtmlWriter::with_options(HtmlWriterOptions {
        strict: true,
        ..Default::default()
    });
    let element = HtmlElement {
        tag: "div!".into(),
        attributes: vec![],
        children: vec![],
        self_closing: true,
    };

    let err = writer.write_html_element(&element).unwrap_err();
    assert!(matches!(err, crate::HtmlWriteError::InvalidHtmlTag(tag) if tag == "div!"));
}

#[test]
fn guarded_writer_textualizes_invalid_attribute_in_non_strict_mode() {
    let mut writer = HtmlWriter::with_options(HtmlWriterOptions {
        strict: false,
        ..Default::default()
    });
    let element = HtmlElement {
        tag: "div".into(),
        attributes: vec![HtmlAttribute {
            name: "onload!".into(),
            value: "evil".into(),
        }],
        children: vec![Node::Text("body".into())],
        self_closing: false,
    };

    writer.write_html_element(&element).unwrap();
    let output = writer.into_string().unwrap();
    assert_eq!(
        output,
        "&lt;div onload!=\"evil\"&gt;\n\nbody\n\n&lt;/div&gt;"
    );
}

#[test]
fn guarded_writer_errors_on_invalid_attribute_in_strict_mode() {
    let mut writer = HtmlWriter::with_options(HtmlWriterOptions {
        strict: true,
        ..Default::default()
    });
    let element = HtmlElement {
        tag: "div".into(),
        attributes: vec![HtmlAttribute {
            name: "onload!".into(),
            value: "evil".into(),
        }],
        children: vec![],
        self_closing: true,
    };

    let err = writer.write_html_element(&element).unwrap_err();
    assert!(matches!(err, crate::HtmlWriteError::InvalidHtmlAttribute(name) if name == "onload!"));
}
