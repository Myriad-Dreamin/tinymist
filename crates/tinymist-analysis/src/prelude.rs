pub use std::ops::Range;
pub use std::path::Path;

pub use ecow::EcoVec;
pub use typst::diag::{EcoString, FileError};
pub use typst::syntax::FileId as TypstFileId;
pub use typst::syntax::{
    ast::{self, AstNode},
    package::{PackageManifest, PackageSpec},
    LinkedNode, Source, SyntaxKind, SyntaxNode, VirtualPath,
};
pub use typst::World;
