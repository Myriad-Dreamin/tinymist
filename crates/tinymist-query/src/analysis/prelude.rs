pub use core::fmt;
pub use std::collections::{BTreeMap, HashMap};
pub use std::hash::{Hash, Hasher};
pub use std::ops::{Deref, Range};
pub use std::path::{Path, PathBuf};
pub use std::sync::{Arc, LazyLock};

pub use comemo::Track;
pub use ecow::*;
pub use serde::{Deserialize, Serialize};
pub use typst::foundations::{Func, Value};
pub use typst::syntax::ast::{self, AstNode};
pub use typst::syntax::{FileId as TypstFileId, LinkedNode, Source, Span, SyntaxKind, SyntaxNode};
pub use typst::World;
pub use typst_shim::syntax::LinkedNodeExt;
pub use typst_shim::utils::LazyHash;

pub use super::AnalysisContext;
pub(crate) use super::StrRef;
pub(crate) use crate::adt::interner::Interned;
pub(crate) use crate::syntax::{IdentRef, LexicalHierarchy, LexicalKind, LexicalModKind, ModSrc};
pub use crate::ty::Ty;
