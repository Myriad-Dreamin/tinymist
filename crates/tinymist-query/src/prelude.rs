pub use std::{
    collections::HashMap,
    iter,
    path::{Path, PathBuf},
    sync::Arc,
};

pub use ecow::eco_vec;
pub use ecow::EcoVec;
pub use itertools::{Format, Itertools};
pub use log::error;
pub use lsp_types::{
    request::GotoDeclarationResponse, CodeAction, CodeActionKind, CodeActionOrCommand, CodeLens,
    ColorInformation, ColorPresentation, CompletionResponse, DiagnosticRelatedInformation,
    DocumentHighlight, DocumentLink, DocumentSymbol, DocumentSymbolResponse, Documentation,
    FoldingRange, GotoDefinitionResponse, Hover, HoverContents, InlayHint, LanguageString,
    Location as LspLocation, LocationLink, MarkedString, MarkupContent, MarkupKind,
    Position as LspPosition, PrepareRenameResponse, SelectionRange, SemanticTokens,
    SemanticTokensDelta, SemanticTokensFullDeltaResult, SemanticTokensResult, SignatureHelp,
    SignatureInformation, SymbolInformation, TextEdit, Url, WorkspaceEdit,
};
pub use reflexo::vector::ir::DefId;
pub use serde_json::Value as JsonValue;
pub use typst::diag::{EcoString, FileResult, Tracepoint};
pub use typst::foundations::Value;
pub use typst::syntax::FileId as TypstFileId;
pub use typst::syntax::{
    ast::{self, AstNode},
    LinkedNode, Source, Spanned, SyntaxKind,
};
pub use typst::World;

pub use crate::analysis::{AnalysisContext, LocalContext};
pub use crate::lsp_typst_boundary::{
    lsp_to_typst, path_to_url, typst_to_lsp, LspDiagnostic, LspRange, LspSeverity,
    PositionEncoding, TypstDiagnostic, TypstSeverity, TypstSpan,
};
pub use crate::{StatefulRequest, VersionedDocument};
