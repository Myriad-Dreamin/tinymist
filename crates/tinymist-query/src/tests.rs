use core::fmt;
use std::{
    collections::{HashMap, HashSet},
    ops::Range,
    path::{Path, PathBuf},
};

use once_cell::sync::Lazy;
pub use serde::Serialize;
use serde_json::{ser::PrettyFormatter, Serializer, Value};
use typst::{
    diag::FileResult,
    syntax::{
        ast::{self, AstNode},
        FileId as TypstFileId, LinkedNode, Source, SyntaxKind, VirtualPath,
    },
};
use typst::{diag::PackageError, foundations::Bytes};
use typst_ts_compiler::{
    service::{CompileDriver, Compiler, EntryManager},
    NotifyApi, ShadowApi,
};
use typst_ts_core::{
    config::compiler::{EntryOpts, EntryState},
    package::Registry,
};
use typst_ts_core::{config::CompileOpts, package::PackageSpec};

pub use insta::assert_snapshot;
pub use typst_ts_compiler::TypstSystemWorld;

use crate::{
    analysis::{Analysis, AnalysisResources},
    prelude::AnalysisContext,
    typst_to_lsp, LspPosition, PositionEncoding,
};

struct WrapWorld<'a>(&'a mut TypstSystemWorld);

impl<'a> AnalysisResources for WrapWorld<'a> {
    fn world(&self) -> &dyn typst::World {
        self.0
    }

    fn resolve(&self, spec: &PackageSpec) -> Result<std::sync::Arc<Path>, PackageError> {
        self.0.registry.resolve(spec)
    }

    fn iter_dependencies<'b>(
        &'b self,
        f: &mut dyn FnMut(&'b reflexo::ImmutPath, FileResult<&typst_ts_compiler::Time>),
    ) {
        self.0.iter_dependencies(f)
    }
}

pub fn snapshot_testing(name: &str, f: &impl Fn(&mut AnalysisContext, PathBuf)) {
    let mut settings = insta::Settings::new();
    settings.set_prepend_module_to_snapshot(false);
    settings.set_snapshot_path(format!("fixtures/{name}/snaps"));
    settings.bind(|| {
        let glob_path = format!("fixtures/{name}/*.typ");
        insta::glob!(&glob_path, |path| {
            let contents = std::fs::read_to_string(path).unwrap();
            #[cfg(windows)]
            let contents = contents.replace("\r\n", "\n");

            run_with_sources(&contents, |w: &mut TypstSystemWorld, p| {
                let root = w.workspace_root().unwrap();
                let paths = w
                    .shadow_paths()
                    .into_iter()
                    .map(|p| {
                        TypstFileId::new(None, VirtualPath::new(p.strip_prefix(&root).unwrap()))
                    })
                    .collect::<Vec<_>>();
                let w = WrapWorld(w);
                let mut ctx = AnalysisContext::new(
                    &w,
                    Analysis {
                        root,
                        position_encoding: PositionEncoding::Utf16,
                        enable_periscope: false,
                        caches: Default::default(),
                    },
                );
                ctx.test_completion_files(Vec::new);
                ctx.test_files(|| paths);
                f(&mut ctx, p);
            });
        });
    });
}

pub fn get_test_properties(s: &str) -> HashMap<&'_ str, &'_ str> {
    let mut props = HashMap::new();
    for line in s.lines() {
        let mut line = line.splitn(2, ':');
        let key = line.next().unwrap().trim();
        let value = line.next().unwrap().trim();
        props.insert(key, value);
    }
    props
}

pub fn run_with_sources<T>(source: &str, f: impl FnOnce(&mut TypstSystemWorld, PathBuf) -> T) -> T {
    let root = if cfg!(windows) {
        PathBuf::from("C:\\")
    } else {
        PathBuf::from("/")
    };
    let mut world = TypstSystemWorld::new(CompileOpts {
        entry: EntryOpts::new_rooted(root.as_path().into(), None),
        ..Default::default()
    })
    .unwrap();
    let sources = source.split("-----");

    let pw = root.join(Path::new("/main.typ"));
    world.map_shadow(&pw, Bytes::from_static(b"")).unwrap();

    let mut last_pw = None;
    for (i, source) in sources.enumerate() {
        // find prelude
        let mut source = source.trim();
        let mut path = None;

        if source.starts_with("//") {
            let first_line = source.lines().next().unwrap();
            let content = first_line.strip_prefix("//").unwrap().trim();

            if let Some(path_attr) = content.strip_prefix("path:") {
                source = source.strip_prefix(first_line).unwrap().trim();
                path = Some(path_attr.trim().to_owned())
            }
        };

        let path = path.unwrap_or_else(|| format!("/s{i}.typ"));

        let pw = root.join(Path::new(&path));
        world
            .map_shadow(&pw, Bytes::from(source.as_bytes()))
            .unwrap();
        last_pw = Some(pw);
    }

    world.mutate_entry(EntryState::new_detached()).unwrap();
    let mut driver = CompileDriver::new(world);
    let _ = driver.compile(&mut Default::default());

    let pw = last_pw.unwrap();
    driver
        .world_mut()
        .mutate_entry(EntryState::new_rooted(
            root.as_path().into(),
            Some(TypstFileId::new(
                None,
                VirtualPath::new(pw.strip_prefix(root).unwrap()),
            )),
        ))
        .unwrap();
    f(driver.world_mut(), pw)
}

pub fn find_test_range(s: &Source) -> Range<usize> {
    // /* range -3..-1 */
    fn find_prefix(s: &str, sub: &str, left: bool) -> Option<(usize, usize, bool)> {
        Some((s.find(sub)?, sub.len(), left))
    }
    let (re_base, re_len, is_after) = find_prefix(s.text(), "/* range after ", true)
        .or_else(|| find_prefix(s.text(), "/* range ", false))
        .unwrap();
    let re_end = re_base + re_len;
    let range_rng = re_end..(s.text()[re_end..].find(" */").unwrap() + re_end);
    let range_base = if is_after {
        range_rng.end + " */".len()
    } else {
        re_base
    };
    let range = &s.text()[range_rng];
    // split by ".."
    let mut bounds = range.split("..");
    // parse the range
    let start: isize = bounds.next().unwrap().parse().unwrap();
    let end: isize = bounds.next().unwrap().parse().unwrap();

    let start = start + range_base as isize;
    let end = end + range_base as isize;
    start as usize..end as usize
}

pub fn find_test_position_after(s: &Source) -> LspPosition {
    find_test_position_(s, 1)
}

pub fn find_test_position(s: &Source) -> LspPosition {
    find_test_position_(s, 0)
}

pub fn find_test_position_(s: &Source, offset: usize) -> LspPosition {
    enum AstMatcher {
        MatchAny { prev: bool },
        MatchIdent { prev: bool },
    }
    use AstMatcher::*;

    let re = s
        .text()
        .find("/* position */")
        .map(|e| (e, MatchAny { prev: true }));
    let re = re.or_else(|| {
        s.text()
            .find("/* position after */")
            .zip(Some(MatchAny { prev: false }))
    });
    let re = re.or_else(|| {
        s.text()
            .find("/* ident */")
            .zip(Some(MatchIdent { prev: true }))
    });
    let re = re.or_else(|| {
        s.text()
            .find("/* ident after */")
            .zip(Some(MatchIdent { prev: false }))
    });
    let (re, m) = re
        .ok_or_else(|| panic!("No position marker found in source:\n{}", s.text()))
        .unwrap();

    let n = LinkedNode::new(s.root());
    let mut n = n.leaf_at(re + 1).unwrap();

    let match_prev = match &m {
        MatchAny { prev } => *prev,
        MatchIdent { prev } => *prev,
    };
    let match_ident = match m {
        MatchAny { .. } => false,
        MatchIdent { .. } => true,
    };

    'match_loop: loop {
        if n.kind().is_trivia() || n.kind().is_error() {
            let m = if match_prev {
                n.prev_sibling()
            } else {
                n.next_sibling()
            };
            n = m.or_else(|| n.parent().cloned()).unwrap();
            continue;
        }
        if matches!(n.kind(), SyntaxKind::Named) {
            n = n.children().last().unwrap();
            continue;
        }
        if match_ident {
            match n.kind() {
                SyntaxKind::Closure => {
                    let c = n.cast::<ast::Closure>().unwrap();
                    if let Some(name) = c.name() {
                        if let Some(m) = n.find(name.span()) {
                            n = m;
                            break 'match_loop;
                        }
                    }
                }
                SyntaxKind::LetBinding => {
                    let c = n.cast::<ast::LetBinding>().unwrap();
                    if let Some(name) = c.kind().bindings().first() {
                        if let Some(m) = n.find(name.span()) {
                            n = m;
                            break 'match_loop;
                        }
                    }
                }
                _ => {}
            }
        }
        break;
    }

    typst_to_lsp::offset_to_position(n.offset() + offset, PositionEncoding::Utf16, s)
}

// pub static REDACT_URI: Lazy<RedactFields> = Lazy::new(||
// RedactFields::from_iter(["uri"]));
pub static REDACT_LOC: Lazy<RedactFields> = Lazy::new(|| {
    RedactFields::from_iter([
        "location",
        "uri",
        "range",
        "selectionRange",
        "targetRange",
        "targetSelectionRange",
        "originSelectionRange",
        "targetUri",
    ])
});

pub struct JsonRepr(Value);

impl JsonRepr {
    pub fn new_pure(v: impl serde::Serialize) -> Self {
        let s = serde_json::to_value(v).unwrap();
        Self(s)
    }

    // pub fn new(v: impl serde::Serialize) -> Self {
    //     let s = serde_json::to_value(v).unwrap();
    //     Self(REDACT_URI.redact(s))
    // }

    pub fn new_redacted(v: impl serde::Serialize, rm: &RedactFields) -> Self {
        let s = serde_json::to_value(v).unwrap();
        Self(rm.redact(s))
    }
}

impl fmt::Display for JsonRepr {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let w = std::io::BufWriter::new(Vec::new());
        let mut ser = Serializer::with_formatter(w, PrettyFormatter::with_indent(b" "));
        self.0.serialize(&mut ser).unwrap();

        f.write_str(&String::from_utf8(ser.into_inner().into_inner().unwrap()).unwrap())
    }
}

pub trait Redact {
    fn redact(&self, v: Value) -> Value;
}

pub struct RedactFields(HashSet<&'static str>);

impl FromIterator<&'static str> for RedactFields {
    fn from_iter<T: IntoIterator<Item = &'static str>>(iter: T) -> Self {
        Self(iter.into_iter().collect())
    }
}

fn pos(v: &Value) -> String {
    match v {
        Value::Object(v) => format!("{}:{}", v["line"], v["character"]),
        Value::Number(v) => v.to_string(),
        _ => "<null>".to_owned(),
    }
}

impl Redact for RedactFields {
    fn redact(&self, v: Value) -> Value {
        match v {
            Value::Object(mut m) => {
                for (_, v) in m.iter_mut() {
                    *v = self.redact(v.clone());
                }
                for k in self.0.iter().copied() {
                    let Some(t) = m.remove(k) else {
                        continue;
                    };

                    match k {
                        "range"
                        | "selectionRange"
                        | "originSelectionRange"
                        | "targetRange"
                        | "targetSelectionRange" => {
                            m.insert(
                                k.to_owned(),
                                format!("{}:{}", pos(&t["start"]), pos(&t["end"])).into(),
                            );
                        }
                        _ => {}
                    }
                }
                Value::Object(m)
            }
            Value::Array(mut a) => {
                for v in a.iter_mut() {
                    *v = self.redact(v.clone());
                }
                Value::Array(a)
            }
            Value::String(s) => Value::String(s),
            v => v,
        }
    }
}

pub struct HashRepr<T>(pub T);

// sha256
impl fmt::Display for HashRepr<JsonRepr> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use sha2::{Digest, Sha256};

        let res = self.0.to_string();
        let hash = Sha256::digest(res).to_vec();
        write!(f, "sha256:{}", hex::encode(hash))
    }
}
