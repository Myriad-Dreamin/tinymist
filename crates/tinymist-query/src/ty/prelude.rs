pub use std::collections::{HashMap, HashSet};

pub use reflexo::vector::ir::DefId;
pub use typst::foundations::Value;

pub use super::builtin::*;
pub use super::def::*;
