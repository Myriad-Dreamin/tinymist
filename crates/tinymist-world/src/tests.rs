use clap::Parser;

use crate::args::CompileOnceArgs;

#[test]
#[cfg(feature = "system")]
fn test_args() {
    let args = CompileOnceArgs::parse_from(["tinymist", "main.typ"]);
    let verse = args
        .resolve_system()
        .expect("failed to resolve system universe");

    let world = verse.snapshot();
    let _res = typst::compile(&world);
}
