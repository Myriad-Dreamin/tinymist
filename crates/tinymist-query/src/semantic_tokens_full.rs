use crate::prelude::*;

/// The [`textDocument/semanticTokens/full`] request is sent from the client to
/// the server to resolve the semantic tokens of a given file.
///
/// [`textDocument/semanticTokens/full`]: https://microsoft.github.io/language-server-protocol/specification#textDocument_semanticTokens
///
/// Semantic tokens are used to add additional color information to a file that
/// depends on language specific symbol information. A semantic token request
/// usually produces a large result. The protocol therefore supports encoding
/// tokens with numbers. In addition, optional support for deltas is available,
/// i.e. [`semantic_tokens_full_delta`].
///
/// [`semantic_tokens_full_delta`]: crate::SemanticTokensDeltaRequest
///
/// # Compatibility
///
/// This request was introduced in specification version 3.16.0.
#[derive(Debug, Clone)]
pub struct SemanticTokensFullRequest {
    /// The path of the document to get semantic tokens for.
    pub path: PathBuf,
}

impl SemanticRequest for SemanticTokensFullRequest {
    type Response = SemanticTokensResult;

    /// Handles the request to compute the semantic tokens for a given document.
    fn request(self, ctx: &mut AnalysisContext) -> Option<Self::Response> {
        let source = ctx.source_by_path(&self.path).ok()?;
        let ei = ctx.expr_stage(&source);
        let token_ctx = &ctx.analysis.tokens_ctx;
        let (tokens, result_id) = token_ctx.semantic_tokens_full(&source, ei);

        Some(
            SemanticTokens {
                result_id: Some(result_id),
                data: tokens,
            }
            .into(),
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tests::*;

    /// This is converted by Copilot from TypeScript `to_multiline_tokens2`.
    /// <https://github.com/microsoft/vscode/blob/2acc0e52cbc7434c415f221d5c34ee1bbdd6cd71/src/vs/editor/common/services/semanticTokensProviderStyling.ts#L147>
    fn check_tokens(tokens: &SemanticTokens) {
        const DESIRED_TOKENS_PER_AREA: usize = 400;
        const DESIRED_MAX_AREAS: usize = 1024;

        let src_data = &tokens.data;
        let token_count = src_data.len();
        let tokens_per_area = std::cmp::max(
            (token_count as f64 / DESIRED_MAX_AREAS as f64).ceil() as usize,
            DESIRED_TOKENS_PER_AREA,
        );

        let mut token_index = 0;
        let mut last_line_number = 1;
        let mut last_start_character = 0;

        while token_index < token_count {
            let token_start_index = token_index;
            let mut token_end_index =
                std::cmp::min(token_start_index + tokens_per_area, token_count);

            // Keep tokens on the same line in the same area...
            if token_end_index < token_count {
                let mut small_token_end_index = token_end_index;
                while small_token_end_index - 1 > token_start_index
                    && src_data[small_token_end_index].delta_line == 0
                {
                    small_token_end_index -= 1;
                }

                if small_token_end_index - 1 == token_start_index {
                    // there are so many tokens on this line that our area would be empty, we must
                    // now go right
                    let mut big_token_end_index = token_end_index;
                    while big_token_end_index + 1 < token_count
                        && src_data[big_token_end_index].delta_line == 0
                    {
                        big_token_end_index += 1;
                    }
                    token_end_index = big_token_end_index;
                } else {
                    token_end_index = small_token_end_index;
                }
            }

            let mut prev_line_number = 0;
            let mut prev_end_character = 0;

            while token_index < token_end_index {
                let delta_line = src_data[token_index].delta_line;
                let delta_character = src_data[token_index].delta_start;
                let length = src_data[token_index].length;
                let line_number = last_line_number + delta_line;
                let start_character = if delta_line == 0 {
                    last_start_character + delta_character
                } else {
                    delta_character
                };
                // delta_character
                let end_character = start_character + length;

                if end_character <= start_character {
                    // this token is invalid (most likely a negative length casted to uint32)
                    panic!(
                        "Invalid length for semantic token at line {line_number}, character {start_character}, end: {end_character}"
                    );
                } else if prev_line_number == line_number && prev_end_character > start_character {
                    // this token overlaps with the previous token
                    panic!("Overlapping semantic tokens at line {line_number}, character {start_character}, previous line {prev_line_number}, previous end {prev_end_character}");
                } else {
                    prev_line_number = line_number;
                    prev_end_character = end_character;
                }

                last_line_number = line_number;
                last_start_character = start_character;
                token_index += 1;
            }
        }
    }

    #[test]
    fn test() {
        snapshot_testing("semantic_tokens", &|ctx, path| {
            let request = SemanticTokensFullRequest { path: path.clone() };

            let mut result = request.request(ctx).unwrap();
            if let SemanticTokensResult::Tokens(tokens) = &mut result {
                tokens.result_id.take();
            }

            match &result {
                SemanticTokensResult::Tokens(tokens) => {
                    check_tokens(tokens);
                }
                SemanticTokensResult::Partial(_) => {
                    panic!("Unexpected partial result");
                }
            }

            assert_snapshot!(serde_json::to_string(&result).unwrap());
        });
    }
}
