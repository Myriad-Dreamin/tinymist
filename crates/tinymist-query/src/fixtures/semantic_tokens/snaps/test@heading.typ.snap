---
source: crates/tinymist-query/src/semantic_tokens_full.rs
expression: "serde_json::to_string(&result).unwrap()"
input_file: crates/tinymist-query/src/fixtures/semantic_tokens/heading.typ
---
{"data":[0,0,3,16,0,0,3,1,22,0,0,1,1,5,0,0,1,4,5,0,0,4,1,10,0,0,1,17,1,0,0,17,1,10,0]}
